Sources/LLBuildManifest/BuildManifest.swift

```diff
@@ -54,6 +56,53 @@ public enum WriteAuxiliary {
         }
     }
 
+    public struct ClangModuleMap: AuxiliaryFileType {
+        public static let name = "modulemap"
+
+        private enum GeneratedModuleMapType: String {
+            case umbrellaDirectory
+            case umbrellaHeader
+        }
+
+        public static func computeInputs(targetName: String, moduleName: String, publicHeadersDir: AbsolutePath, type: PackageLoading.GeneratedModuleMapType) -> [Node] {
+            let typeNodes: [Node]
+            switch type {
+            case .umbrellaDirectory(let path):
+                typeNodes = [.virtual(GeneratedModuleMapType.umbrellaDirectory.rawValue), .directory(path)]
+            case .umbrellaHeader(let path):
+                typeNodes = [.virtual(GeneratedModuleMapType.umbrellaHeader.rawValue), .file(path)]
+            }
+            return [.virtual(Self.name), .virtual(targetName), .virtual(moduleName), .directory(publicHeadersDir)] + typeNodes
+        }
+
+        public static func getFileContents(inputs: [Node]) throws -> String {
+            guard inputs.count == 5 else {
+                throw StringError("invalid module map generation task, inputs: \(inputs)")
+            }
+
+            let generator = ModuleMapGenerator(
+                targetName: inputs[0].extractedVirtualNodeName,
+                moduleName: inputs[1].extractedVirtualNodeName,
+                publicHeadersDir: try AbsolutePath(validating: inputs[2].name),
+                fileSystem: localFileSystem
+            )
+
+            let declaredType = inputs[3].extractedVirtualNodeName
+            let path = try AbsolutePath(validating: inputs[4].name)
+            let type: PackageLoading.GeneratedModuleMapType
+            switch declaredType {
+            case GeneratedModuleMapType.umbrellaDirectory.rawValue:
+                type = .umbrellaDirectory(path)
+            case GeneratedModuleMapType.umbrellaHeader.rawValue:
+                type = .umbrellaHeader(path)
+            default:
+                throw StringError("invalid module map type in generation task: \(declaredType)")
```

Comment:
Could this use a specific error enum case instead of a stringly-typed error? That would make it easier to assert on this exact error when testing an unhappy path of this function.

Reply:
I don't think that's necessary in this case since these are largely theoretical programming errors that shouldn't ever occur in practice.